nemo/core/callbacks.py

```diff
@@ -35,6 +54,463 @@
+                tensors_logged['LR'] = state["learning_rate"]
+                self._wandb_log(tensors_logged)
+
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self._wandb_log({"epoch": state["epoch"], "epoch_time": epoch_time})
+
+    def _wandb_log(self, tensors_logged):
+        if _WANDB_AVAILABLE:
+            wandb.log(tensors_logged, step=state["step"])
+
+
+class CheckpointCallback(NeMoCallback):
+    def __init__(
+        self,
+        folder: str,
+        load_from_folder: str = None,
+        step_freq: int = -1,
+        epoch_freq: int = -1,
+        checkpoints_to_keep: int = 4,
+        force_load: bool = False,
+    ):
+        """A callback that does checkpointing of module weights and trainer (incl. optimizer) status.
+
+        args:
+            folder (str, required): A path where checkpoints are to be stored and loaded from if load_from_folder is
+                None.
+            load_from_folder (str): A path where checkpoints can be loaded from.
+                Defaults to None.
+            step_freq (int): How often in terms of steps to save checkpoints. One of step_freq or epoch_freq is
+                required.
+            epoch_freq (int): How often in terms of epochs to save checkpoints. One of step_freq or epoch_freq is
+                required.
+            checkpoints_to_keep (int): Number of most recent checkpoints to keep. Older checkpoints will be deleted.
+                Defaults to 4.
+            force_load (bool): Whether to crash if loading is unsuccessful.
+                Defaults to False
+        """
+        if step_freq == -1 and epoch_freq == -1:
+            logging.warning("No checkpoints will be saved because step_freq and epoch_freq are both -1.")
+
+        if step_freq > -1 and epoch_freq > -1:
+            logging.warning("You config the model to save by both steps and epochs. Please use one or the other")
+            epoch_freq = -1
+
+        self._step_freq = step_freq
+        self._epoch_freq = epoch_freq
+        self._folder = folder
+        self._load_from_folder = load_from_folder if load_from_folder else folder
+        self._ckpt2keep = checkpoints_to_keep
+        self._saved_ckpts = []
+        # If True, run will fail if we cannot load module weights
+        self._force_load = force_load
+
+    def __save_to(self, path, state):
+        if state["global_rank"] is not None and state["global_rank"] != 0:
+            return
+        if not os.path.isdir(path):
+            logging.info(f"Creating {path} folder")
+            os.makedirs(path, exist_ok=True)
+        unique_mod_names = set()
+        for module in AppState().modules:
```

Comment:
I truly believe this callback should work in the scope of _a single graph_.

Why are we saving all the modules?